client-s3/src/main/kotlin/app/cash/backfila/client/s3/internal/S3DatasourceBackfillOperator.kt

```diff
@@ -0,0 +1,184 @@
+import app.cash.backfila.protos.clientservice.GetNextBatchRangeResponse
+import app.cash.backfila.protos.clientservice.GetNextBatchRangeResponse.Batch
+import app.cash.backfila.protos.clientservice.KeyRange
+import app.cash.backfila.protos.clientservice.PrepareBackfillRequest
+import app.cash.backfila.protos.clientservice.PrepareBackfillResponse
+import app.cash.backfila.protos.clientservice.RunBatchRequest
+import app.cash.backfila.protos.clientservice.RunBatchResponse
+import com.google.common.base.Stopwatch
+import java.util.concurrent.TimeUnit
+import okio.ByteString.Companion.encodeUtf8
+
+class S3DatasourceBackfillOperator<R : Any, P : Any>(
+  override val backfill: S3DatasourceBackfill<R, P>,
+  private val parametersOperator: BackfilaParametersOperator<P>,
+  private val s3Service: S3Service,
+) : BackfillOperator {
+
+  override fun name(): String = backfill.javaClass.toString()
+
+  override fun prepareBackfill(request: PrepareBackfillRequest): PrepareBackfillResponse {
+    val config = parametersOperator.constructBackfillConfig(request)
+    backfill.validate(config)
+
+    require(request.range?.start == null && request.range?.end == null) {
+      // We could think about supporting this later by making the range mean a byte seek into all S3 files.
+      // This would mean we would need to support some kind of seek forward or seek back for record start.
+      // That or perhaps we only support it for single file?
+      // In any case, we are not implementing this now.
+      "Range is currently unsupported for S3 Backfils"
+    }
+
+    val pathPrefix = backfill.getPrefix(config)
+    val fileKeys = s3Service.listFiles(backfill.getBucket(config), pathPrefix)
+
+    // Sanity check that this backfill will actually process something
+    require(fileKeys.isNotEmpty()) {
+      "No files found for bucket:${backfill.getBucket(config)} prefix:$pathPrefix. At least one file must exist."
+    }
+    require(fileKeys.size <= 100) {
+      "Listing files matching the prefix contains ${fileKeys.size} which is more than 100 files. " +
+        "Check your prefix. First 3 file keys ${fileKeys.slice(0..2)}"
+    }
+
+    // The path after the prefix must be no more than 45 characters.
+    val postfixes = fileKeys.map { it.removePrefix(pathPrefix) }
+    val invalidPostfixes = postfixes.filter { it.length > 45 }
+    require(invalidPostfixes.isEmpty()) {
+      "Found invalid postfixes: $invalidPostfixes"
+    }
+
+    val partitions = postfixes.map {
+      PrepareBackfillResponse.Partition.Builder()
+        .partition_name(it)
+        .backfill_range(
+          KeyRange.Builder()
+            .start((0L).toString().encodeUtf8())
+            .build(),
+        )
+        .build()
+    }
+    return PrepareBackfillResponse.Builder()
+      .partitions(partitions)
+      .build()
+  }
+
+  override fun getNextBatchRange(request: GetNextBatchRangeRequest): GetNextBatchRangeResponse {
```

Comment:
nit: this method is super long, is there a way to clean it up a bit by extracting some private methods?

Reply:
I'll do this after